### PR TITLE
fix(types): make compatible with Vue 2.5 and TypeScript 2.6

### LINKED ISSUE #272: Types test broken in Vue 2.5.x
`npm run test:types` fails on Vue 2.5.x. 

There are quite a few errors thrown. The types changed significantly in Vue 2.5.x. We need to fix the broken tests so they compile using types from the latest Vue version

### PR DESCRIPTION
**Pull request by ktsn**
fix #272

There are two reasons that the types test is broken:

1. The 1st type parameter of `ComponentOptions` is no longer specify the `this` type of lifecycle hooks because Vue 2.5 uses the intersection type with `ThisType` to do that.
2. [TypeScript 2.6 had introduced more strict type checking for function types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html). So the `Component` type cannot pass the type checking.

I've added `ThisType`d version of `MountOptions` and `ShallowOptions` to solve 1. Also added a new `Component` type to solve 2.

### CODE REVIEW

**Comment by ktsn** [comment]
I'm not sure why the unit test failed 😕 
Maybe bumping to Vue 2.5 affects the behavior of `find`?

File: types/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,9 @@
-import Vue, { VNodeData, Component, ComponentOptions, FunctionalComponentOptions } from 'vue'
+import Vue, { VNodeData, ComponentOptions, FunctionalComponentOptions } from 'vue'
+
+export type Component =
```
Comment (HerringtonDarkholme) [inline types/index.d.ts]: If `Component` type from vue package fails, it probably means we need to fix it in the main repo. Would you like to point out the failing usage?

Comment (ktsn) [reply types/index.d.ts]: Indeed. I found the `Component` type always has the base `Vue` type for `this` type of `Data` and `Method` in default. That fails type checking if `ComponentOptions` has different type: https://github.com/vuejs/vue-test-utils/blob/dev/types/test/resources.ts#L9

Maybe we should change the default type parameters of `Component` (and `AsyncComponent`) to `{}`?

Comment (eddyerburgh) [reply types/index.d.ts]: Shall we merge this now and update the Component type here later when it's fixed in the main repo?

Comment (HerringtonDarkholme) [reply types/index.d.ts]: Let's merge it now. We can change it later after new Vue is released.

Comment (HerringtonDarkholme) [reply types/index.d.ts]: @ktsn The problem is that `Component` isn't correctly setting up type bound. 

`ComponentOption<V>` is contravariant, that is, `ComponentOption<Normal>` isn't a sub type of `ComponentOption<Vue>`, but of `ComponentOption<never>`.

`beforeCreate` in componentOption would also break assignment here if contravariant method was checked. But currently it is not https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#note

Comment (ktsn) [reply types/index.d.ts]: @HerringtonDarkholme Thanks for clarify the problem! I didn't realize we can use `never` for the default generic type.

I'll change the `ComponentOptions`'s 1st type parameter to `never` for now before this PR is merged.

**Review by eddyerburgh — APPROVED** [review]

**Pull request merged**